### DISCUSSION TITLE
Integration with Iron Bird

### DISCUSSION THREAD

**Discussion by sthelia** [JSBSim-Team/jsbsim#393] [Support requests]
Hi everyone! 
I’m an aerospace student and have just started my master thesis work. The project is to try to integrate JSBSim with an Iron Bird for testing control surface actuators. 

I’m currently in the start up phase, getting to know the program and the rig I’ll be working on, so I’m sure questions will arise as time goes on. Happy to get input of what would be interesting to include in the project or things to consider before getting started! 

My current idea is that I want to continuously pass some properties, say dynamic pressure and requested elevator deflection, to the rig. Later I would also like to integrate the actual deflection as an input to JSBSim. The rig is controlled through Simulink (and speedgoat). I’ve seen that JSBSim is capable of both input and output of different types. 

I’m not familiar with using telnet, but would that be a good option to looking into if JSBSim and Simulink ran on different computers? Would there be a difference if they’re running on separate computers and connected by cable? 

Say that I would like to run a simple example, for example the simple ball example. If I want to run it in realtime, I add the `--realtime` when running the program. If I would like to see the altitude updated at 10Hz, I'm thinking I would do something similar to: 

``` xml
<output name="localhost" type="SOCKET" port="????"  rate="10">
	<property> position/h-sl-meters </property>
</output>
```

What does the name `localhost` signify in this case? How do I determine which port number to use? Can I see the output in another terminal if I access the port with `telnet localhost port` or something similar? I'm on Mac. 

Thanks in advance and have a great weekend!

**Comment by bcoconni** [comment]
Thanks for using JSBSim for your master thesis :+1:
> What does the name `localhost` signify in this case? How do I determine which port number to use? Can I see the output in another terminal if I access the port with `telnet localhost port` or something similar? I'm on Mac.

Here `localhost` is the host name of the computer you want to interface with. You can supply an IP address as well.

Port is the TCP/UDP port number and can be any number that complies with the OSes of the 2 computers that will communicate (see https://en.wikipedia.org/wiki/List_of_TCP_and_UDP_port_numbers for example of valid port numbers).

**Reply by sthelia** [reply]
Thanks! 
I managed to get some real time data in the terminal by using netcat, `nc -l -p PORT`, and then initialising my JSBSim file in realtime. Still struggling to get it to Matlab, but it's a start! I've found Matlab contains both TCP and UDP functions, but I haven't managed to get it to read data in real time.

**Comment by seanmcleod** [comment]
Out of interest, does your Iron Bird rig consist of hardware from some specific aircraft, or are they just some generic actuators?

**Reply by sthelia** [reply]
It's the empennage of Saab 2000, however there are some modifications of the actuators going on simultaneously. Linköping University is now researching electric actuators, so these are not original anymore. The loading system is also being upgraded and I hope this will allow simulations for other aircraft, such as fighters. 

Some previous thesis work on the Iron Bird for those interested: 
[http://liu.diva-portal.org/smash/get/diva2:234610/FULLTEXT01.pdf](http://liu.diva-portal.org/smash/get/diva2:234610/FULLTEXT01.pdf )

**Comment by seanmcleod** [comment]
@sthelia thanks for background info, always useful and interesting to know how JSBSim is being used.

**Comment by sthelia** [comment]
Is there any information about how the data is being sent over socket via TCP or UDP? 

My current workaround just to see what is going on with Matlab is to create a TCP server, then start the JSBSim running. Matlab then looks at BytesAvailable and reads that. The data is received as unicode and I can then translate that to a string. This is however not real time, I can read around 2 seconds worth of data here. 
However, I can't tell if there's a convenient way to split the data to figure out which fields are time and which are properties. I did see a "new line" command that I now manually used to split the data. 

I'm trying to make sense of what a package, byte, datagram etc is and how JSBSim handles this? 

I also just read this on Wikipedia, which might make me want to do UDP instead since the real time compatibility is going to be important. 

> TCP is optimized for accurate delivery rather than timely delivery and can incur relatively long delays (on the order of seconds) while waiting for out-of-order messages or re-transmissions of lost messages. Therefore, it is not particularly suitable for real-time applications such as voice over IP. For such applications, protocols like the Real-time Transport Protocol (RTP) operating over the User Datagram Protocol (UDP) are usually recommended instead.

**Comment by seanmcleod** [comment]
> Is there any information about how the data is being sent over socket

The most accurate and up to date information is the source code :wink:

There is a JSBSim Reference Manual although it hasn't been updated in a long time, but the section on socket output is still pretty accurate.

http://jsbsim.sourceforge.net/JSBSimReferenceManual.pdf

First thing to be aware of is that there are two different socket output options that you can specify.

```xml
<output type="SOCKET"..... >

<output type="FLIGHTGEAR"..... >
```

The main difference between the two is that the `FLIGHTGEAR` option sends a fixed binary packet, designed specifically for driving FlightGear although that doesn't preclude you from using it assuming it includes all the required data fields you need, versus the `SOCKET` option which sends the data in plain text (pretty sure it's plain ASCII as opposed to Unicode) but has the advantage that you get to specify the exact set of properties that you want sent.

Depending on the tools/programming language you're using to receive the network packets reading the plain text packets may be easier than processing a binary structure.

You mentioned running netcat successfully earlier right, so that shows you exactly the plain text data being sent via the `SOCKET` option, e.g. I added the following to the 737 cruise script example.

```xml
  <output name="localhost" type="SOCKET" port="1138" rate="1">
        <property>velocities/vc-kts</property>
        <property>position/h-agl-ft</property>
  </output>
```

You can see that data looks very similar (excluding the `<LABELS>` string to csv data.

```sh
nc.exe -l -p 1138
<LABELS> ,Time,vc-kts,h-agl-ft
           0,    285.0745,       30000
     0.99996,    283.9128,    29993.57
     1.99992,    283.0492,    29974.78
     2.99988,    282.5345,    29941.87
```

So in general you would need to split the incoming data based on a newline, then split it based on commas and then you would have the individual fields as plain text versions of the properties you selected to transmit.

At this stage I wouldn't worry about the difference between TCP and UDP. You're going to be running this on a LAN right, it's not as if the one PC is going to be sitting somewhere on the internet with multiple routers between the PCs.
